api/state.go

```diff
@@ -128,24 +221,64 @@ func (st *state) Login(tag names.Tag, password, nonce string, macaroons []macaro
 				// they presented was invalid.
 				err = cause.(*httpbakery.InteractionError).Reason
 			}
-			return errors.Trace(err)
+			return result, errors.Trace(err)
 		}
 		// Add the macaroons that have been saved by HandleError to our login request.
 		request.Macaroons = httpbakery.MacaroonsForURL(st.bakeryClient.Client.Jar, st.cookieURL)
 		result = params.LoginResult{} // zero result
 		err = st.APICall("Admin", 3, "", "Login", request, &result)
 		if err != nil {
-			return errors.Trace(err)
+			return result, errors.Trace(err)
 		}
 		if result.DischargeRequired != nil {
-			return errors.Errorf("login with discharged macaroons failed: %s", result.DischargeRequiredReason)
+			return result, errors.Errorf("login with discharged macaroons failed: %s", result.DischargeRequiredReason)
+		}
+	}
+	return result, nil
+}
+
+// Login authenticates as the entity with the given name and password
+// or macaroons. Subsequent requests on the state will act as that entity.
+// This method is usually called automatically by Open. The machine nonce
+// should be empty unless logging in as a machine agent.
+func (st *state) Login(p LoginParams) error {
+	var result params.LoginResult
+	var err error
+
+	switch {
+	case p.AccessToken == "" && p.ClientID == "" && p.ClientSecret == "" && p.Password == "" && p.Macaroons == nil:
```

Comment:
Following on from previous comment, I think this is messy and we are trying to stuff orthogonal use cases into the one Login call. What about having public login methods for

LoginWithUserCredentials <- the current v3 login
LoginWithClientCredentials
LoginWithDeviceFlow
LoginWithAccessToken

or something